realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -413,38 +414,47 @@ private static void initializeRealm(Realm realm) {
         // the Realm is initialized.
         boolean commitChanges = false;
         try {
-            realm.beginTransaction();
+            // We need to start a transaction no matter readOnly mode, because it acts as an interprocess lock.
+            // TODO: For proper inter-process support we also need to move e.g copying the asset file under an
+            // interprocess lock. This lock can obviously not be created by a Realm instance so we probably need
+            // to implement it in Object Store. When this happens, the `beginTransaction(true)` can be removed again.
+            realm.beginTransaction(true);
```

Comment:
I agree, this needs to be changed with the OS schema integration. Which was why I went this route as it seemed the approach with the lowest friction.

It is not true that we cannot prevent multi-process interactions here. This is the entire reason the setup/validate method is inside a write-transaction. That is cross-process safe. If another process then later changes the Schema, the SharedRealm in this process will recieve a schema change callback.

The downside of course is that you need to know _not_ to write anything inside this method.
Having an interprocess lock that doesn't depend on Realm would be safer, and also required for our `assetFile()` to work. I don't think Object Store has this kind of logic yet. Nor do I know how difficult it is to add. 

Java doesn't officially support interprocess yet though, so your point is valid. In read-only mode, just open the Realm and verify. Then we can tackle interprocess later? That would also address @bmeike issues around code complexity in this method.